src/vse_sync_pp/analyzers/pmc.py

```diff
@@ -0,0 +1,112 @@
+"""Analyze PMC log messages"""
+
+from .analyzer import Analyzer
+
+STATE_FREERUN = 248
+STATE_LOCKED = 6
+STATE_HOLDOVER_IN_SPEC = 7
+STATE_HOLDOVER_OUT_OF_SPEC1 = 140
+STATE_HOLDOVER_OUT_OF_SPEC2 = 150
+STATE_HOLDOVER_OUT_OF_SPEC3 = 160
+
+STATE_TRANSITION = {
+    STATE_FREERUN: [STATE_FREERUN, STATE_LOCKED],
+    STATE_LOCKED: [STATE_LOCKED, STATE_HOLDOVER_IN_SPEC],
+    STATE_HOLDOVER_IN_SPEC: [STATE_LOCKED, 
+                             STATE_HOLDOVER_IN_SPEC, 
+                             STATE_HOLDOVER_OUT_OF_SPEC1, 
+                             STATE_HOLDOVER_OUT_OF_SPEC2, 
+                             STATE_HOLDOVER_OUT_OF_SPEC3],
+    STATE_HOLDOVER_OUT_OF_SPEC1: [STATE_LOCKED, 
+                                  STATE_HOLDOVER_OUT_OF_SPEC1, 
+                                  STATE_HOLDOVER_OUT_OF_SPEC2, 
+                                  STATE_HOLDOVER_OUT_OF_SPEC3],
+    STATE_HOLDOVER_OUT_OF_SPEC2: [STATE_LOCKED, 
+                                  STATE_HOLDOVER_OUT_OF_SPEC1, 
+                                  STATE_HOLDOVER_OUT_OF_SPEC2, 
+                                  STATE_HOLDOVER_OUT_OF_SPEC3],
+    STATE_HOLDOVER_OUT_OF_SPEC3: [STATE_LOCKED, 
+                                  STATE_HOLDOVER_OUT_OF_SPEC1, 
+                                  STATE_HOLDOVER_OUT_OF_SPEC2, 
+                                  STATE_HOLDOVER_OUT_OF_SPEC3],
+}
+
+def is_transition_allowed(current_state, new_state):
+   return new_state in STATE_TRANSITION[current_state]
+
+class ClockStateAnalyzer(Analyzer):
+    """Analyze clock state
+    """
+    id_ = 'phc/gm-settings'
+    parser = id_
+    def __init__(self, config):
+        super().__init__(config)
+        # minimum test duration for a valid test
+        self._duration = config.parameter('min-test-duration/s')
+        self.state = None
+        self.transition_count = 0
+        self.clock_class_count = {
+            STATE_FREERUN: 0,
+            STATE_LOCKED: 0,
+            STATE_HOLDOVER_IN_SPEC: 0,
+            STATE_HOLDOVER_OUT_OF_SPEC1: 0,
+            STATE_HOLDOVER_OUT_OF_SPEC2: 0,
+            STATE_HOLDOVER_OUT_OF_SPEC3: 0,
+        }
+    def prepare(self, rows):
+        idx = 0
+        try:
+            tstart = rows[0].timestamp
+        except IndexError:
+            pass
+        else:
+            while idx < len(rows):
+                if tstart <= rows[idx].timestamp:
+                    break
+                idx += 1
+        return super().prepare(rows[idx:])
+    def test(self, data):
+        if len(data) == 0:
+            return (False, "no data")
+        
+        for index, row in data.iterrows():
+            clock_class = row['clock_class']
+ 
+            if (self.state is None) and (clock_class in STATE_TRANSITION):
+                self.state = clock_class
+            else:
+                if clock_class != self.state:
+                    self.transition_count += 1
+                
+                if clock_class in STATE_TRANSITION:
```

Comment:
Because the if after this uses `transition_allowed` its essentially an extension of the true block. I would invert this if so it reads something like

```python
    if clock_class not in STATE_TRANSITIONS:
        return (False, "unknown clock class")

    if is_transition_allowed(self.state, clock_class):
        self.state = clock_class
        self.clock_class_count[clock_class] += 1
    else:
        return (False, "state transition not allowed")
```

Reply:
done